backend/Dockerfile

```diff
@@ -1,7 +1,10 @@
 ##################################################################################
 # BASE ###########################################################################
 ##################################################################################
-FROM node:18.7.0-alpine3.16 as base
+# base on custom docker container which is essential node:18.7.0-alpine3.16 
+# but with installed build-tools for rust native npm modules
+# and with compiled @iota/client
+FROM gradido/iota_alpine:2.2.4 as base
```

Comment:
this must be checked in - where is this built?

Reply:
On my PC, uploaded to hub.docker.com.
Da du immer sehr spezifische Vorstellungen hast von sowas, sag du mir wo genau soll ich das Dockerfile hintun?
Ein eigenes Repository oder hier mit rein? In einen Unterordner?

Reply:
Packs einfach in dieses Image wie im ersten Review von mir vorgeschlagen - also so wie dus gemacht hast ohne Zwischenlayer.

Der Grund für meine "sehr spezifische Vorstellungen" ist einfach eine zusätzlich Abhängigkeit von einem unbekannten Dockerfile - alle Abhängigkeiten sollten teil des Projekts sein oder zumindest in einem Repo verfügbar. Das gilt für alle unsere Abhängigkeiten - auch die externen. Damit zu brechen und das ohne Grund macht keinen Sinn.

Aber lange Rede keinen Sinn: bitte baue es direkt ins image ohne zwischen layer wie in meinem ersten Review angemerkt. Wenn du Hilfe brauchst sag bescheid.

Reply:
Ich glaube du hast nicht genau verstanden warum ich das ausgelagert habe.
Beim builden des ausgelagerten Docker-Files wird der @iota/client compiliert, also das was die Buildzeit erhöht.
Wenn ich das direkt in das Dockerfile vom backend reinschreibe, dann haben wir wie zuvor, eine massive erhöhte Buildzeit. 

Den Zwischenlayer habe ich gemacht, damit die zusätzlichen 1GB die durch das installieren der Rust Build-tools anfallen nicht im Production Layer landen. 
Da steht im Kommentar: PRODUCTION (Does contain only "binary"- and static-files to reduce image size)
Wäre irgendwie doof wenn ich das dann wieder durch die rust build tools aufblähe, oder?
Zum Vergleich unkomprimiert
- master branch backend production build: 697 MB
- iota3_bootstrap_hello_welt branch backend production build: 3,19 GB
Die aktuelle Lösung berücksichtigt deinen Änderungswunsch ohne Extra Layer auszukommen. 

Und falls dir das nicht klar ist, die Rust Build Tools brauche ich nur zum compilieren von @iota/client nicht zum ausführen.
Das kompilieren erzeugt eine Binär Datei namens index.node die sich im Ordner node_modules/@iota/client/build/Release befindet und etwa 10 MB groß ist. Die enthält den ganzen in Rust geschriebenen iota client compiliert in Maschinensprache. Das brauche ich zum ausführen, mehr nicht.

Reply:
@iota/client ist ein Wrapper für den iota_client der in Rust geschrieben ist.
yarn add @iota/client benötigt die rust Build-Tools wenn es kein prebuild Binary für das verwendete System gibt. 
Versuche es mal auf deinem System, dann siehst du was ich meine.

Reply:
es gibt auch eine ältere reine javascript implementierung mit der Alex im Frontend gerade rumspielt.
Das habe ich auch schon probiert, leider bekomme ich dann beim Testen immer folgende Fehlermeldung:
```bash
Jest has detected the following 2 open handles potentially keeping Jest from exiting:

  ●  TLSWRAP

      57 |     return null
      58 |   }
    > 59 |   return sendData(
         |          ^
      60 |     client,
      61 |     Converter.utf8ToBytes(CONFIG.IOTA_COMMUNITY_ALIAS),
      62 |     Converter.utf8ToBytes(message),

      at node_modules/node-fetch/lib/index.js:1468:15
      at fetch (node_modules/node-fetch/lib/index.js:1437:9)
      at SingleNodeClient.fetchWithTimeout (node_modules/@iota/iota.js/dist/cjs/index-node.js:1815:40)
      at SingleNodeClient.fetchJson (node_modules/@iota/iota.js/dist/cjs/index-node.js:1706:41)
      at SingleNodeClient.messageSubmit (node_modules/@iota/iota.js/dist/cjs/index-node.js:1511:41)
      at sendData (node_modules/@iota/iota.js/dist/cjs/index-node.js:2625:45)
      at sendDataMessage (src/apis/IotaConnector.ts:59:10)
      at src/apis/IotaConnector.test.ts:19:48
      at src/apis/IotaConnector.test.ts:8:71
      at Object.<anonymous>.__awaiter (src/apis/IotaConnector.test.ts:4:12)
      at Object.<anonymous> (src/apis/IotaConnector.test.ts:18:63)
      at TestScheduler.scheduleTests (node_modules/@jest/core/build/TestScheduler.js:333:13)
      at runJest (node_modules/@jest/core/build/runJest.js:387:19)
      at _run10000 (node_modules/@jest/core/build/cli/index.js:320:7)
      at runCLI (node_modules/@jest/core/build/cli/index.js:173:3)


  ●  TLSWRAP

      74 |     return null
      75 |   }
    > 76 |   return client.message(messageId)
         |                 ^
      77 | }
      78 |
      79 | /**

      at node_modules/node-fetch/lib/index.js:1468:15
      at fetch (node_modules/node-fetch/lib/index.js:1437:9)
      at SingleNodeClient.fetchWithTimeout (node_modules/@iota/iota.js/dist/cjs/index-node.js:1815:40)
      at SingleNodeClient.fetchJson (node_modules/@iota/iota.js/dist/cjs/index-node.js:1706:41)
      at SingleNodeClient.message (node_modules/@iota/iota.js/dist/cjs/index-node.js:1462:25)
      at getMessage (src/apis/IotaConnector.ts:76:17)
      at src/apis/IotaConnector.test.ts:32:43
      at src/apis/IotaConnector.test.ts:8:71
      at Object.<anonymous>.__awaiter (src/apis/IotaConnector.test.ts:4:12)
      at Object.<anonymous> (src/apis/IotaConnector.test.ts:29:82)
      at TestScheduler.scheduleTests (node_modules/@jest/core/build/TestScheduler.js:333:13)
      at runJest (node_modules/@jest/core/build/runJest.js:387:19)
      at _run10000 (node_modules/@jest/core/build/cli/index.js:320:7)
      at runCLI (node_modules/@jest/core/build/cli/index.js:173:3)
```